Remove False Alarm Non-Breaking Whitespace

jojo2357 opened the pull request:
Fix #3122

![image](https://github.com/Hannah-Sten/TeXiFy-IDEA/assets/66704796/1ebf9250-8998-4eb6-b9c8-f0463fefbf8d)

all tests passed locally

PHPirates reviewed (changes requested):
Thanks for looking into this! I don't quite agree to this though, I think we should keep the warning when there is a normal space (that is, in the output pdf) instead of a non-breaking one in the middle of sentence. I added a test to highlight this.

jojo2357 commented:
I think that should do it

PHPirates reviewed (approved):
Indeed this does fix the original issue, so that's great, but if you _only_ check if there is no sentence end, you are missing all the other cases in which the \ref is not in the middle of a sentence. For example,

```
    \begin{center}
        Something centered.
    \end{center}
    \ref[blub} Starts a new sentence.
```
is a false positive now.
On the other hand,
```This sentence has a $\xi$ \ref{in} it.```
is a true positive.

I don't see an easy way out of this, do you? If not, let's go with this.

Maybe we can go for providing a second quickfix in case the previous element is not a LatexNormalText (and the next word starts with a capital letter) which suggests that starting a sentence with a reference is bad style. We do it like that for impossible cases in the LatexLineBreakInspection as well.

jojo2357 commented:
So I added your tests and some of my own. I also added a simple check for the second previous token being an actual text token but that is only temporary. Some wilder changes are on their way, I just made this commit so that when you hate the cringe im about to do you can easily save the pr :P

jojo2357 commented:
What to do about ``$yes.$ \ref{} please.`` ? Warn or not?

PHPirates commented:
> What to do about `$yes.$ \ref{} please.` ? Warn or not?

The full stop is inside the inline math so it's math and does not end a sentence. If it does, I would say it's bad style to put it inside the math. However, for display math this is less clear, because you clearly don't want to put the full stop on its own line.
I don't think this is an important case though, let's not spend too much time on that.

jojo2357 commented:
Well i fixed a lot of cases so unless you can find any other major ones, im calling it good. 

Heres one: suppose `\textit` was like this: `\textit{Heres the text.}{italics}` then i think this will not behave correctly, however I know not a command that does this.

PHPirates reviewed (approved):
Yeah there is always going to be an infinite amount of ways to fool the inspection, also for the Grazie implementation we have the trouble of not knowing which characters are actual text and thus which ones to send to Grazie for grammar checking. But this PR is already a huge improvement, awesome!

jojo2357 commented:
I agree. We can improve it by request rather than hold up all progress whilst awaiting perfection

PHPirates commented:
You perfectly summed up my approach to TeXiFy in general, I just wait for issues to come in 😆
